model/job-management/src/main/proto/beam_expansion_api.proto

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Protocol Buffers describing the Expansion API, an api for expanding
+ * transforms in a remote SDK.
+ */
+
+syntax = "proto3";
+
+package org.apache.beam.model.expansion.v1;
+
+option go_package = "construction_v1";
+option java_package = "org.apache.beam.model.expansion.v1";
+option java_outer_classname = "ExpansionApi";
+
+import "beam_runner_api.proto";
+
+message ExpansionRequest {
+  // Set of components needed to interpret the transform, or which
+  // may be useful for its expansion.  This includes the input
+  // PCollections (if any) to the to-be-expanded transform, along
+  // with their coders and windowing strategies.
+  org.apache.beam.model.pipeline.v1.Components components = 1;
+
+  // The actual PTransform to be expaneded according to its spec.
+  // Its input should be set, but its subtransforms and outputs
+  // should not be.
+  org.apache.beam.model.pipeline.v1.PTransform transform = 2;
+
+  // A namespace (prefix) to use for the id of any newly created
+  // components.
+  string namespace = 3;
```

Comment:
Worth noting that this is not necessary. The whole idea of components is that they never have meaning outside the message they are contained in, and are just a way of encoding a graph. It is fair to critique that the copying and re-allocating ids has computational cost. The last time we relied on ids having significance outside a single message we hit bugs and migration pain. Those are expected consequences IMO.

Here, it seems that it is not so bad. IIUC you are just using this namespace as a way to cheaply reuse components, but not establishing any cross-message dependence upon them.

But it is worth noting that you can do a role-reversal and save yourself this field: the expansion service can freely prefix all the _other_ ids with something to avoid collisions with its chosen namespace.

Reply:
Ah, I had forgotten about the opaque payloads. That's a good point. We could address this (for example by having Components elsewhere than just the top level) but I don't think we should. All is good here.